org.osgi.test.junit5/src/main/java/org/osgi/test/junit5/service/ServiceExtension.java

```diff
@@ -114,31 +113,27 @@ public boolean supportsParameter(ParameterContext parameterContext, ExtensionCon
 
 		if (List.class.equals(memberType) && (genericMemberType instanceof ParameterizedType)) {
 			serviceType = ((ParameterizedType) genericMemberType).getActualTypeArguments()[0];
-		}
-		else if (ServiceAware.class.equals(memberType) && (genericMemberType instanceof ParameterizedType)) {
+		} else if (ServiceAware.class.equals(memberType) && (genericMemberType instanceof ParameterizedType)) {
 			serviceType = ((ParameterizedType) genericMemberType).getActualTypeArguments()[0];
 		}
 		// The service type must be a raw type
 		if (serviceType instanceof Class) {
 			return true;
 		}
-
-		return false;
+		throw new ParameterResolutionException("Can only resolve @" + InjectService.class.getSimpleName()
+			+ " parameter for services with non-generic types, service type was: " + serviceType.getTypeName());
 	}
 
 	static void assertValidFieldCandidate(Field field) {
 		if (Modifier.isFinal(field.getModifiers()) || Modifier.isPrivate(field.getModifiers())
 			|| Modifier.isStatic(field.getModifiers())) {
-			throw new ExtensionConfigurationException(
-				InjectService.class.getName() + " field [" + field + "] must not be final, private or static.");
+			throw new ExtensionConfigurationException('@' + InjectService.class.getSimpleName() + " field ["
```

Comment:
`"@"` since we are concatenating strings, we should use String constants. Not a char constant which must be converted to a String object at runtime.

Reply:
Fixed.

I didn't believe it until I looked at the decompiled output - I had assumed that the compiler would generate a call `StringBuilder.append(char)` rather than `StringBuilder.append(String.valueOf(char))`. Why doesn't it???

Reply:
> Why doesn't it???

🤷‍♂️